### PR TITLE
System

### LINKED ISSUE #50: As a framework user I want to define and use Systems
**Initial estimate of effort:** 5

#### Sprint tasks
- [x] #51
- [x] #52

### LINKED ISSUE #51: Define a way to create systems

### LINKED ISSUE #52: Add systems to the world

### LINKED ISSUE #53: As a framework user I want to update the World's state
**Initial estimate of effort:** 5

### PR DESCRIPTION
**Pull request by giacomocavalieri**
Closes #50 
Closes #51 
Closes #52 
Closes #53

### CODE REVIEW

File: core/src/main/scala/dev/atedeg/ecscala/System.scala [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+import scala.annotation.tailrec
+import dev.atedeg.ecscala.{ CList, Entity, View }
+import dev.atedeg.ecscala.util.types.given
+import dev.atedeg.ecscala.util.types.{ CListTag, ComponentTag }
+
+trait System[L <: CList] extends Function3[Entity, L, View[L], Deletable[L]] {
+  def before(view: View[L]): Unit = {}
+
+  def after(view: View[L]): Unit = {}
```
Comment (ndido98) [inline core/src/main/scala/dev/atedeg/ecscala/System.scala:11]: Add a Scaladoc for these two methods

File: core/src/main/scala/dev/atedeg/ecscala/System.scala [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+  // This method should only be called with a CListTag[L], it can not be enforced in the interface as
+  // it would make it impossible to call it correctly from the World requiring to cast the tags to a
+  // type that has been erased.
```
Comment (ndido98) [inline core/src/main/scala/dev/atedeg/ecscala/System.scala:15]: Make it a Scaladoc?

File: core/src/main/scala/dev/atedeg/ecscala/dsl/EcscalaDSL.scala [hunk 3]
```diff
@@ -1,6 +1,6 @@
 package dev.atedeg.ecscala.dsl
```
Comment (ndido98) [inline core/src/main/scala/dev/atedeg/ecscala/dsl/EcscalaDSL.scala:1]: Rename the file as ECScalaDSL.scala

File: core/src/main/scala/dev/atedeg/ecscala/util/macros/ViewMacro.scala [hunk 4]
```diff
@@ -25,29 +26,28 @@ object ViewMacro {
...
-      case _: (head &: tail) =>
-        world.getComponents[head].get(entity) &: getEntityComponents[tail](world)(entity)
-      case _ => CNil
-    }).asInstanceOf[L]
+  private def getEntityComponents[L <: CList](
+      taggedMaps: Seq[(ComponentTag[? <: Component], Map[Entity, ? <: Component])],
+  )(entity: Entity)(using clt: CListTag[L]): L = {
+    val taggedComponents = taggedMaps
+      .map(taggedMap => taggedMap._1 -> taggedMap._2(entity))
+      .asInstanceOf[Seq[(ComponentTag[Component], Component)]]
```
Comment (ndido98) [inline core/src/main/scala/dev/atedeg/ecscala/util/macros/ViewMacro.scala:40]: Why is this cast safe? Explain in a comment

File: core/src/main/scala/dev/atedeg/ecscala/util/macros/ViewMacro.scala [hunk 5]
```diff
@@ -25,29 +26,28 @@ object ViewMacro {
...
-    (inline erasedValue[L] match {
-      case _: (head &: tail) =>
-        world.getComponents[head].get(entity) &: getEntityComponents[tail](world)(entity)
-      case _ => CNil
-    }).asInstanceOf[L]
+  private def getEntityComponents[L <: CList](
+      taggedMaps: Seq[(ComponentTag[? <: Component], Map[Entity, ? <: Component])],
+  )(entity: Entity)(using clt: CListTag[L]): L = {
+    val taggedComponents = taggedMaps
+      .map(taggedMap => taggedMap._1 -> taggedMap._2(entity))
```
Comment (ndido98) [inline core/src/main/scala/dev/atedeg/ecscala/util/macros/ViewMacro.scala:39]: Remove the dot and parentheses for map

File: core/src/main/scala/dev/atedeg/ecscala/util/macros/ViewMacro.scala [hunk 6]
```diff
@@ -25,29 +26,28 @@ object ViewMacro {
...
+      .asInstanceOf[Seq[(ComponentTag[Component], Component)]]
+    taggedComponents
+      .foldRight(CNil: CList)((compTag, acc) => &:(compTag._2, acc)(using compTag._1))
+      .asInstanceOf[L]
```
Comment (ndido98) [inline core/src/main/scala/dev/atedeg/ecscala/util/macros/ViewMacro.scala:43]: Why is this cast safe? Explain in a comment

File: core/src/test/scala/dev/atedeg/ecscala/dsl/EcscalaDSLTest.scala [hunk 7]
```diff
@@ -2,7 +2,7 @@ package dev.atedeg.ecscala.dsl
```
Comment (ndido98) [inline core/src/test/scala/dev/atedeg/ecscala/dsl/EcscalaDSLTest.scala:2]: Rename file in ECScalaDSLTest.scala

**Review by ndido98 — CHANGES_REQUESTED** [review]

**Review by nicolasfara — APPROVED** [review]

File: core/src/main/scala/dev/atedeg/ecscala/System.scala (hunk 1 above)
Comment (giacomocavalieri) [reply core/src/main/scala/dev/atedeg/ecscala/System.scala:11]: Why would you need a `World` parameter here? Should I add it to the `update` method too if I'm adding it?

File: core/src/main/scala/dev/atedeg/ecscala/System.scala (hunk 2 above)
Comment (giacomocavalieri) [reply core/src/main/scala/dev/atedeg/ecscala/System.scala:15]: Even if the method is `private[ecscala]`?

File: core/src/main/scala/dev/atedeg/ecscala/System.scala (hunk 1 above)
Comment (ndido98) [reply core/src/main/scala/dev/atedeg/ecscala/System.scala:11]: The world parameter could be used if I need to add components/entities in those before/after methods, and we already have it in the update method

File: core/src/main/scala/dev/atedeg/ecscala/System.scala (hunk 2 above)
Comment (ndido98) [reply core/src/main/scala/dev/atedeg/ecscala/System.scala:15]: Yes, so it appears correctly using an IDE

**Review by vitlinda — APPROVED** [review]

File: core/src/main/scala/dev/atedeg/ecscala/util/macros/ViewMacro.scala (hunk 5 above)
Comment (nicolasfara) [reply core/src/main/scala/dev/atedeg/ecscala/util/macros/ViewMacro.scala:39]: For symmetry with `foldRight` use the `.` also in `map`

**Review by ndido98 — APPROVED** [review]

**Review by ndido98 — APPROVED** [review]

**Pull request merged**